### PR TITLE
mysql (ticdc): fix batch dml delete sql gennerate error

### LINKED ISSUE #9812: TiCDC did not correctly use the primary key as the where clause for the Delete statement.
### What did you do?

I use TiCDC to synchronize a table with one primary key (consisting of two columns of int values) and a non-null unique index (consisting of 3 columns). I observed from the logs that the where statement of the Delete statement generated by TiCDC did not correctly use the primary key as an execution condition.

### What did you expect to see?

TiCDC uses primary key to generate the where clause of delete statements.

### What did you see instead?

as above.

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
(paste TiDB cluster version here)
```

Upstream TiKV version (execute `tikv-server --version`):

```console
(paste TiKV version here)
```

TiCDC version (execute `cdc version`):

```console
(paste TiCDC version here)
```

### PR DESCRIPTION
**Pull request by asddongmen**
### What problem does this PR solve?

Issue Number: close #9812 

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/sink/dmlsink/txn/mysql/mysql.go [hunk 1]
```diff
@@ -567,8 +567,15 @@ func (s *mysqlBackend) prepareDMLs() *preparedDMLs {
 			}
 			// only use batch dml when the table has a handle key
 			if hasHandleKey(tableColumns) {
-				// TODO(dongmen): find a better way to get table info.
-				tableInfo := model.BuildTiDBTableInfo(tableColumns, firstRow.IndexColumns)
+				var tableInfo *timodel.TableInfo
+				if firstRow.TableInfo != nil && firstRow.TableInfo.TableInfo != nil {
+					tableInfo = firstRow.TableInfo.TableInfo
+				} else {
+					// firstRow.TableInfo is nil only when the row change event was generated by
```
Comment (asddongmen) [inline cdc/sink/dmlsink/txn/mysql/mysql.go:574]:
```suggestion
					// firstRow.TableInfo is nil only when the row change event was **not** generated by
```

**Review by nongfushanquan — APPROVED** [review]

File: cdc/sink/dmlsink/txn/mysql/mysql.go [hunk 2]
```diff
@@ -567,8 +567,15 @@ func (s *mysqlBackend) prepareDMLs() *preparedDMLs {
...
-				// TODO(dongmen): find a better way to get table info.
-				tableInfo := model.BuildTiDBTableInfo(tableColumns, firstRow.IndexColumns)
+				var tableInfo *timodel.TableInfo
+				if firstRow.TableInfo != nil && firstRow.TableInfo.TableInfo != nil {
+					tableInfo = firstRow.TableInfo.TableInfo
+				} else {
+					// firstRow.TableInfo is nil only when the row change event was **not** generated by
+					// mounter, which only occurs when mysql sink was used in redo consumer or
+					// in the other consumers.
+					tableInfo = model.BuildTiDBTableInfo(tableColumns, firstRow.IndexColumns)
```
Comment (sdojjy) [inline cdc/sink/dmlsink/txn/mysql/mysql.go:577]: `BuildTiDBTableInfo` generates incorrect index information, it's better to fix it or add a comment

File: cdc/entry/mounter_test.go [hunk 3]
```diff
@@ -1615,7 +1653,18 @@ func TestBuildTableInfo(t *testing.T) {
 		handle := sqlmodel.GetWhereHandle(recoveredTI, recoveredTI)
 		require.NotNil(t, handle.UniqueNotNullIdx)
 		require.Equal(t, c.recovered, showCreateTable(t, recoveredTI))
-
+		// make sure BuildTiDBTableInfo indentify the correct primary key
+		if c.caseNumber == 5 {
```
Comment (lance6716) [inline cdc/entry/mounter_test.go:1657]: caseNumber can be replaced by `i` of `for i, c := range cases {`

File: cdc/model/sink.go [hunk 4]
```diff
@@ -604,7 +604,10 @@ func BuildTiDBTableInfo(columns []*Column, indexColumns [][]int) *model.TableInf
 			continue
 		}
 		if firstCol.Flag.IsPrimaryKey() {
```
Comment (lance6716) [inline cdc/model/sink.go:606]: maybe we can postpone the recover of PK in line 617's loop

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: cdc/model/sink.go (hunk 4 above)
Comment (asddongmen) [reply cdc/model/sink.go:606]: Thank you for your advice!

File: cdc/entry/mounter_test.go (hunk 3 above)
Comment (asddongmen) [reply cdc/entry/mounter_test.go]: sure

**Review by lance6716 — APPROVED** [review]

**Pull request merged**